Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp

```diff
@@ -681,57 +675,91 @@ void ABenchmarkGymGameModeBase::OnRepTotalNPCs()
 	OnTotalNPCsUpdated(TotalNPCs);
 }
 
-int32 ABenchmarkGymGameModeBase::GetActorClassCount(TSubclassOf<AActor> ActorClass) const
+void ABenchmarkGymGameModeBase::OnActorCountReportId()
 {
-	const UCounterComponent* CounterComponent = Cast<UCounterComponent>(GameState->GetComponentByClass(UCounterComponent::StaticClass()));
-	check(CounterComponent != nullptr)
-
-	return CounterComponent->GetActorClassCount(ActorClass);
+	UpdateAndReportActorCounts();
 }
 
-void ABenchmarkGymGameModeBase::SetLifetime(int32 Lifetime)
+void ABenchmarkGymGameModeBase::UpdateAndReportActorCounts()
 {
-	if (TestLifetimeTimer.SetTimer(Lifetime))
+	const UWorld* World = GetWorld();
+	const FString WorkerID = GetGameInstance()->GetSpatialWorkerId();
+	if (WorkerID.IsEmpty())
 	{
-		TestLifetimeTimer.SetLock(true);
-		UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("Setting NFR test lifetime %d"), Lifetime);
+		return;
 	}
-	else
+
+	ActorCountMap& ThisWorkerActorCounts = WorkerActorCounts.FindOrAdd(WorkerID);
+	for (auto const& Pair : ExpectedActorCounts)
 	{
-		UE_LOG(LogBenchmarkGymGameModeBase, Warning, TEXT("Could not set NFR test liftime to %d. Timer was locked."), Lifetime);
+		const TSubclassOf<AActor>& ActorClass = Pair.Key;
+		if (!USpatialStatics::IsActorGroupOwnerForClass(World, ActorClass))
+		{
+			continue;
+		}
+
+		const FExpectedActorCountConfig& Config = Pair.Value;
+		if (Config.MinCount > 0)
+		{
+			ThisWorkerActorCounts.FindOrAdd(ActorClass) = GetActorAuthCount(ActorClass);
+		}
+	}
+
+	TArray<FActorCount> ActorCountArray;
+	ActorCountArray.Reserve(ThisWorkerActorCounts.Num());
+
+	for (const auto& Pair : ThisWorkerActorCounts)
+	{
+		ActorCountArray.Add(FActorCount(Pair.Key, Pair.Value));
 	}
+
+	ReportAuthoritativeActorCount(ActorCountReportId, WorkerID, ActorCountArray);
 }
 
-void ABenchmarkGymGameModeBase::ReportAuthoritativePlayers_Implementation(const FString& WorkerID, const int AuthoritativePlayers)
+int32 ABenchmarkGymGameModeBase::GetActorAuthCount(const TSubclassOf<AActor>& ActorClass) const
 {
-	if (HasAuthority())
+	const UWorld* World = GetWorld();
+	USpatialNetDriver* SpatialDriver = Cast<USpatialNetDriver>(World->GetNetDriver());
+	USpatialPackageMapClient* PackageMap = SpatialDriver->PackageMap;
+	const SpatialGDK::EntityView& View = SpatialDriver->Connection->GetView();
+
+	TArray<AActor*> Actors;
+	UGameplayStatics::GetAllActorsOfClass(World, ActorClass, Actors);
+
+	int32 AuthCount = 0;
+	for (const AActor* Actor : Actors)
 	{
-		MapAuthoritativePlayers.Emplace(WorkerID, AuthoritativePlayers);
-		int32 TotalPlayers = 0;
-		for (const auto& KeyValue : MapAuthoritativePlayers)
+		if (Actor->HasAuthority())
 		{
-			TotalPlayers += KeyValue.Value;
+			AuthCount++;
 		}
-		if (MapAuthoritativePlayers.Num() == NumWorkers)
+		else if (PackageMap != nullptr)
```

Comment:
Should we `ensure` this?

Reply:
Nah, but we defo need to make it work in native :/